cuqi/sampler/_gibbs.py

```diff
@@ -0,0 +1,224 @@
+        A, y_obs, probinfo = cuqi.testproblem.Deconvolution1D.get_components(phantom='square')
+        n = A.domain_dim
+
+        # Define distributions
+        d = cuqi.distribution.Gamma(1, 1e-4)
+        l = cuqi.distribution.Gamma(1, 1e-4)
+        x = cuqi.distribution.GMRF(np.zeros(n), lambda d: d[0])
+        y = cuqi.distribution.GaussianCov(A, lambda l: 1/l)
+
+        # Combine into a joint distribution and create posterior
+        joint = cuqi.distribution.JointDistribution([d, l, x, y])
+        posterior = joint(y=y_obs)
+
+        # Define sampling strategy
+        sampling_strategy = {
+            'x': cuqi.sampler.Linear_RTO,
+            ('d', 'l'): cuqi.sampler.Conjugate,
+        }
+
+        # Define Gibbs sampler
+        sampler = cuqi.sampler.Gibbs(posterior, sampling_strategy)
+
+        # Run sampler
+        samples = sampler.sample(Ns=1000, Nb=200)
+
+        # Plot results
+        samples['x'].plot_ci(exact=probinfo.exactSolution)
+        samples['d'].plot_trace(figsize=(8,2))
+        samples['l'].plot_trace(figsize=(8,2))
+            
+    """
+
+    def __init__(self, target: JointDistribution, sampling_strategy: Dict[Union[str,tuple], Sampler]):
+
+        # Store target and allow conditioning to reduce to a single density
+        self.target = target
+        self.target._allow_reduce = True
+
+        # Parse samplers and split any keys that are tuple into separate keys
+        self.samplers = {}
+        for par_name in sampling_strategy.keys():
+            if isinstance(par_name, tuple):
+                for par_name_ in par_name:
+                    self.samplers[par_name_] = sampling_strategy[par_name]
+            else:
+                self.samplers[par_name] = sampling_strategy[par_name]
+
+        # Store parameter names
+        self.par_names = self.target.get_parameter_names()
+
+    # ------------ Public methods ------------
+    def sample(self, Ns, Nb=0):
+        """ Sample from target distribution """
+
+        # Initial points
+        current_samples = self._get_initial_points()
+
+        # Compute how many samples were already taken previously
+        at_Nb = self._Nb
+        at_Ns = self._Ns
+
+        # Allocate memory for samples
+        self._allocate_samples_warmup(Nb)
+        self._allocate_samples(Ns)
+
+        # Sample tuning phase
+        for i in range(at_Nb, at_Nb+Nb):
+            current_samples = self.step_tune(current_samples)
+            self._store_samples(self.samples_warmup, current_samples, i)
+            self._print_progress(i+1+at_Nb, at_Nb+Nb, 'Warmup')
+
+        # Sample phase
+        for i in range(at_Ns, at_Ns+Ns):
+            current_samples = self.step(current_samples)
+            self._store_samples(self.samples, current_samples, i)
+            self._print_progress(i+1, at_Ns+Ns, 'Sample')
+
+        # Convert to samples objects and return
+        return self._convert_to_Samples(self.samples)
+
+    def step(self, current_samples):
```

Comment:
Does it need to be passed as input and returned at the end? Would be an idea to do `self.current_samples` instead?

Reply:
I think we should leave this until we decide on a good structure for the base Sampler class #34.

The Gibbs sampler would subclass from Sampler